common/rpc/interceptor/namespace_validator.go

```diff
@@ -0,0 +1,142 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package interceptor
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/workflowservice/v1"
+	"google.golang.org/grpc"
+
+	"go.temporal.io/server/common/namespace"
+)
+
+type (
+	// NamespaceValidatorInterceptor validates:
+	// 1. Namespace is specified in request if there is a `namespace` field.
+	// 2. Namespace exists.
+	// 3. Namespace is in correct state.
+	NamespaceValidatorInterceptor struct {
+		namespaceRegistry namespace.Registry
+	}
+)
+
+var (
+	ErrNamespaceNotSet              = serviceerror.NewInvalidArgument("Namespace not set on request.")
+	errInvalidNamespaceStateMessage = "Namespace has invalid state: %s. Must be %s."
+
+	// By default, enumspb.NAMESPACE_STATE_REGISTERED and enumspb.NAMESPACE_STATE_DEPRECATED are allowed states for all APIs that have `namespace` field in the request object.
+	allowedNamespaceStates = map[string][]enumspb.NamespaceState{
+		"StartWorkflowExecution":           {enumspb.NAMESPACE_STATE_REGISTERED},
+		"SignalWithStartWorkflowExecution": {enumspb.NAMESPACE_STATE_REGISTERED},
+		"DescribeNamespace":                {enumspb.NAMESPACE_STATE_REGISTERED, enumspb.NAMESPACE_STATE_DEPRECATED, enumspb.NAMESPACE_STATE_DELETED},
+	}
+)
+
+var _ grpc.UnaryServerInterceptor = (*NamespaceValidatorInterceptor)(nil).Intercept
+
+func NewNamespaceValidatorInterceptor(
+	namespaceRegistry namespace.Registry,
+) *NamespaceValidatorInterceptor {
+	return &NamespaceValidatorInterceptor{
+		namespaceRegistry: namespaceRegistry,
+	}
+}
+
+func (ni *NamespaceValidatorInterceptor) Intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	namespaceEntry, err := ni.extractNamespace(req)
+	if err != nil {
+		return nil, err
+	}
+	if namespaceEntry != nil {
+		_, methodName := splitMethodName(info.FullMethod)
+		allowedStates, allowedStatesDefined := allowedNamespaceStates[methodName]
+		if !allowedStatesDefined {
+			// If not explicitly defined, only enumspb.NAMESPACE_STATE_REGISTERED and enumspb.NAMESPACE_STATE_DEPRECATED are allowed.
+			if namespaceEntry.State() != enumspb.NAMESPACE_STATE_REGISTERED && namespaceEntry.State() != enumspb.NAMESPACE_STATE_DEPRECATED {
+				return nil, serviceerror.NewInvalidArgument(fmt.Sprintf(errInvalidNamespaceStateMessage, namespaceEntry.State(), fmt.Sprintf("%s or %s", enumspb.NAMESPACE_STATE_REGISTERED.String(), enumspb.NAMESPACE_STATE_DEPRECATED.String())))
+			}
```

Comment:
I would just set allowedStates = defaultAllowedStates here, and remove the else. Also define as a const.